### PR TITLE
DM-34964: add cancel, restart and ping function in the bps panda plugin

### PR DESCRIPTION
**Pull request by wguanicedew**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 1]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+                    "jobs": [],
+                    "job_state_counts": job_state_counts,
+                }
+
+                wms_report = WmsRunReport(**report)
+                run_reports.append(wms_report)
+                success = True
+
+        if not success:
+            raise RuntimeError(f"Error to get workflow status: {str(ret)}")
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:425]: Again, `str` not needed.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 2]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+            (e.g. PanDA) can safely ignore it.
+
+        Returns
+        -------
+        req_ids : `list` [`Any`]
+            Only job ids to be used by cancel and other functions.  Typically
+            this means top-level jobs (i.e., not children jobs).
+        """
+        c = self.get_idds_client()
+        ret = c.get_requests(request_id=wms_id)
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:461]: Better variable name needed.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 3]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+        _LOG.debug("PanDA get workflows returned = %s", str(ret))
+
+        req_ids = []
+        success = False
+        if ret[0] == 0 and ret[1][0]:
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:466]: This logic is used a lot and it is starting to worry me that `ret` should be a class on which you can ask questions like this.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 4]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+            reqs = ret[1][1]
+            for req in reqs:
+                req_ids.append(req["request_id"])
+                success = True
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:470]: Setting `success` to `True` every time round the loop seems excessive. I think you are saying that `success` should be `True` if `reqs` has any content? How about:

```python
req_ids = []
if ret[0] == 0 and ret[1][0]:
    req_ids = [req["request_id"] for req in ret[1][1]]

if not req_ids:
    raise RuntimeError(...)

return req_ids
```

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 5]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+        -------
+        deleted : `bool`
+            Whether successful deletion or not.  Currently, if any doubt or any
+            individual jobs not deleted, return False.
+        message : `str`
+            Any message from WMS (e.g., error details).
+        """
+        c = self.get_idds_client()
+        ret = c.abort(request_id=wms_id)
+        _LOG.debug("Abort PanDA workflow returned = %s", str(ret))
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:495]: No `str` needed

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 6]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+        Returns
+        -------
+        deleted : `bool`
+            Whether successful deletion or not.  Currently, if any doubt or any
+            individual jobs not deleted, return False.
+        message : `str`
+            Any message from WMS (e.g., error details).
+        """
+        c = self.get_idds_client()
+        ret = c.abort(request_id=wms_id)
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:494]: variable name.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 7]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+        _LOG.debug("Abort PanDA workflow returned = %s", str(ret))
+
+        rets = []
+        success = True
+        if ret[0] == 0 and ret[1][0]:
+            for req in ret[1][1]:
+                if type(req) in [tuple, list]:
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:501]: Restrictive type check. At least comment that you are deliberately constraining to tuple/list and no duck-typing a Sequence.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 8]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+            individual jobs not deleted, return False.
+        message : `str`
+            Any message from WMS (e.g., error details).
+        """
+        c = self.get_idds_client()
+        ret = c.abort(request_id=wms_id)
+        _LOG.debug("Abort PanDA workflow returned = %s", str(ret))
+
+        rets = []
+        success = True
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:498]: How does `success` ever get set to `False`?

**Review by timj — CHANGES_REQUESTED** [review]
I am assuming @MichelleGower will also take a look from a bps perspective.

I have many comments but I think most of them are easy to deal with.

I do have a concern about the return values from idds APIs being very hard to use. It looks like there are tuples with tuples and the code here would benefit significantly from those APIs returning objects that can be inspected with methods.

Code like this:

```
if ret[0] == 0 and ret[1][0]
```
is very hard to follow and would be much easier if `ret` was an object that could have methods run on it.

File: doc/changes/DM-34964.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+Add cancel,restart, report and ping functions in bps panda plugin.
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply doc/changes/DM-34964.rst:1]: fixed.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 10]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
 
         return files_plc_hldr, direct_IO_files
 
+    def get_idds_client(self):
+        """Get the idds client
+        Returns
+        -------
+        idds_client: `iDDS client manager object`
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:276]: fixed.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 11]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+        """
+        idds_server = None
+        if type(self.config) in [BpsConfig]:
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:279]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 6 above)
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:494]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 12]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+            _, idds_server = self.config.search("iddsServer", opt={"default": None})
+        elif type(self.config) in [dict] and "iddsServer" in self.config:
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:281]: I updated to use isinstance.
For bps submit, a bps yaml file is read. So it returns BpsConfig and a config.search function can be called.
But in bps abort, for example, no yaml file is read. Here the config part is a dict. It doesn't have search function.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 13]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+            idds_server = self.config["iddsServer"]
+        idds_client = pandaclient.idds_api.get_api(
+            idds_utils.json_dumps, idds_host=idds_server, compress=True, manager=True
+        )
+        return idds_client
+
+    def restart(self, wms_workflow_id):
+        """Restart a workflow from the point of failure.
+        Parameters
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:290]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 14]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+                        _LOG.info(req[1])
+                        rets.append(req[1])
+                    else:
+                        _LOG.error(req[1])
+                        rets.append(req[1])
+                else:
+                    _LOG.info(req)
+                    rets.append(req)
+        if not success:
+            raise RuntimeError(f"Error to retry workflow: {str(ret)}")
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:327]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 15]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
         message = ""
-        run_reports = None
+        run_reports = []
+
+        c = self.get_idds_client()
+        ret = c.get_requests(request_id=wms_workflow_id, with_detail=True)
+        _LOG.debug("PanDA get workflow status returned = %s", str(ret))
+
+        success = False
+        if ret[0] == 0 and ret[1][0]:
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:393]: new functions to replace it.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 16]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+            reqs = ret[1][1]
+            for req in reqs:
+                transform_status = req["transform_status"]["attributes"]["_name_"]
+                wms_state = self.convert_idds_state_to_wms_state(transform_status)
+                job_state_counts = {}
+                for state in WmsStates:
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:399]: It's required in bps. Without setting this part, bps report will raise errors.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 4 above)
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:470]: fixed.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/ctrl/bps/panda/panda_service.py:399]: I'm confused. I was suggesting a more compact way of filling the `job_state_counts` variable. Not saying to remove the code. It looked like the three lines could be replaced by the one I suggested.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 17]
```diff
@@ -297,9 +383,156 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
             pointers to documentation or to WMS specific commands.
         """
         message = ""
-        run_reports = None
+        run_reports = []
+
+        c = self.get_idds_client()
+        ret = c.get_requests(request_id=wms_workflow_id, with_detail=True)
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:389]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 18]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+            to None.
+        run_name : `str`
+            Name of the restarted workflow. If restart failed, it will be set
+            to None.
+        message : `str`
+            A message describing any issues encountered during the restart.
+            If there were no issue, an empty string is returned.
+        """
+        c = self.get_idds_client()
+        ret = c.retry(request_id=wms_workflow_id)
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:308]: yes. However I reformat this function.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 19]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+        _LOG.debug("Retry PanDA workflow returned = %s", str(ret))
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:309]: The ret is a tuple. without str, only ret[0] will match the '%s'. Then it will raise errors such as not all items are formatted.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 20]
```diff
@@ -269,6 +269,92 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+
+        rets = []
+        success = True
+        if ret[0] == 0 and ret[1][0]:
+
+            for req in ret[1][1]:
+                if type(req) in [tuple, list]:
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:316]: reformatted this function.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/ctrl/bps/panda/panda_service.py:309]: I don't think so:

```
>>> x = (1, 3, 4)
>>> logging.warning("Tuple: %s", x)
WARNING:root:Tuple: (1, 3, 4)
```
That seems to stringify the tuple (you aren't doing `*ret`).

[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:353]: ok. updated. Thanks.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:435]: Sorry. I missed understood. Thanks.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 21]
```diff
@@ -269,6 +268,126 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+            idds_utils.json_dumps, idds_host=idds_server, compress=True, manager=True
+        )
+        return idds_client
+
+    def get_idds_result(self, ret):
+        """Parse the results returned from iDDS.
+
+        Parameters
+        ----------
+            ret: `(int, (bool, payload))`
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:296]: The indent is wrong her and below.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 22]
```diff
@@ -269,6 +268,126 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
 
         return files_plc_hldr, direct_IO_files
 
+    def get_idds_client(self):
+        """Get the idds client
+
+        Returns
+        -------
+        idds_client: `ClientManager`
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:276]: Does this render okay if you do `package-docs build`? I think it might need the full name of the iDDS class.

Comment (timj) [reply python/lsst/ctrl/bps/panda/panda_service.py:275]: @wguanicedew change the target python version in pyproject.toml. We forgot to change it to py310 when we migrated to py310.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 23]
```diff
@@ -297,8 +415,141 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+                    "run_summary": None,
+                    "state": wms_state,
+                    "total_number_jobs": req["output_total_files"],
+                    "jobs": [],
+                    "job_state_counts": job_state_counts,
+                }
+
+                wms_report = WmsRunReport(**report)
+                run_reports.append(wms_report)
+            return run_reports, message
```
Comment (MichelleGower) [inline python/lsst/ctrl/bps/panda/panda_service.py:459]: When I run this with `--id 1512` I get 21 runs.  I expected 1 run with 120+ jobs across 21 job labels.

**Review by MichelleGower — COMMENTED** [review]
In my tests, report isn't working as expected.   Also, the behavior in error cases isn't the same as the expected behavior for all of the new functions.  And a heads up that the expected ping method signature will be different than it appears here.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 24]
```diff
@@ -269,6 +267,126 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+        idds_client = self.get_idds_client()
+        ret = idds_client.retry(request_id=wms_workflow_id)
+        _LOG.debug("Restart PanDA workflow returned = %s", ret)
+
+        status, result, error = self.get_idds_result(ret)
+        if status:
+            _LOG.info("Restarting PanDA workflow %s", result)
+            return wms_workflow_id, None, json.dumps(result)
+        else:
+            raise RuntimeError(f"Error retry PanDA workflow: {error}")
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:359]: Working to reformat the functions to replace RuntimeError to messages. Will test them again.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 23 above)
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:459]: Here I put one run per panda task. What do you mean job labels? How to map 21 panda tasks to different job labels?

Comment (MichelleGower) [reply python/lsst/ctrl/bps/panda/panda_service.py:459]: a bps "run" is a single butler output collection which is a single workflow being submitted to WMS.   As for how to map PanDA tasks to job labels, I'm not sure how to do this in PanDA.  Each GenericWorkflowJob has a label.  I don't know if that's lost once submitted to PanDA or not.

Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:459]: ok. I will check at first whether the label is kept or not. Then we can discuss how to map the jobs to different labels.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 21 above)
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:295]: updated and tested with "package-docs build"

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 25]
```diff
@@ -269,6 +267,126 @@ def copy_files_for_distribution(tasks, file_distribution_uri):
...
+    def convert_idds_state_to_wms_state(self, state):
+        """Convert iDDS state to BPS wms state
+        Parameters
+        ----------
+        state : `str`:
+            iDDS task state.
+
+        Returns
+        -------
+        wms_state: `WmsStates`
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:370]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 26]
```diff
@@ -297,8 +415,141 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
         """
         message = ""
-        run_reports = None
-        return run_reports, message
+        run_reports = []
+
+        idds_client = self.get_idds_client()
+        ret_requests = idds_client.get_requests(request_id=wms_workflow_id, with_detail=True)
+        _LOG.debug("PanDA get workflow status returned = %s", ret_requests)
+
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:423]: updated to add lines to report errors.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 22 above)
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py:275]: updated and tested ok with packages-docs build.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 27]
```diff
@@ -297,8 +415,141 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+        _LOG.debug("Abort PanDA workflow returned = %s", ret)
+
+        status, result, error = self.get_idds_result(ret)
+        if status:
+            _LOG.info("Aborting PanDA workflow %s", result)
+            return True, json.dumps(result)
+        else:
+            raise RuntimeError(f"Error abort PanDA workflow: {error}")
+
+    def ping(self):
```
[the remark this answers is not in the document]
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py]: updated.

File: python/lsst/ctrl/bps/panda/panda_service.py (hunk 23 above)
Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py]: report function is removed from this ticket.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 28]
```diff
@@ -296,9 +385,116 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+
+        Returns
+        -------
+        req_ids : `list` [`Any`]
+            Only job ids to be used by cancel and other functions.  Typically
+            this means top-level jobs (i.e., not children jobs).
+        """
+        if wms_id is None and user is not None:
+            msg = "Error to get workflow status report: wms_id is required"
+            msg += " and filtering worflows with 'user' is not supported."
```
Comment (MichelleGower) [inline python/lsst/ctrl/bps/panda/panda_service.py:423]: Thanks for adding the check and nicer error message.

Typo (missing k):  filtering worflows -> filtering workflows

Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py]: fixed.

File: python/lsst/ctrl/bps/panda/panda_service.py [hunk 29]
```diff
@@ -296,9 +385,116 @@ def report(self, wms_workflow_id=None, user=None, hist=0, pass_thru=None, is_glo
...
+        Returns
+        -------
+        req_ids : `list` [`Any`]
+            Only job ids to be used by cancel and other functions.  Typically
+            this means top-level jobs (i.e., not children jobs).
+        """
+        if wms_id is None and user is not None:
+            msg = "Error to get workflow status report: wms_id is required"
+            msg += " and filtering workflows with 'user' is not supported."
+            raise RuntimeError(msg)
```
Comment (timj) [inline python/lsst/ctrl/bps/panda/panda_service.py:424]: Can you do it instead as:
```python
raise RuntimeError("Error to get workflow status report: wms_id is required"
    " and filtering workflows with 'user' is not supported.")
```
there is no real need to do the separate variable and concatenation. I think the `msg` line could be done without the `+=`:

```python
msg = ("Error to get workflow status report: wms_id is required"
       " and filtering workflows with 'user' is not supported.")
```
(and black would likely fix it up for you if you just made the line long)

Comment (wguanicedew) [reply python/lsst/ctrl/bps/panda/panda_service.py]: It's good. Thanks. I updated it.
